File: src/lib/adaptive.ts
import type { DifficultyLevel, Skill } from "@/lib/types";
import { getAllSkills } from "@/lib/exam/curriculum";

// ─── Types ────────────────────────────────────────────────────────────

export type ConfidenceTrend = "improving" | "stable" | "declining";

export type SessionMode = "teach" | "practice";

export interface StudentSkillState {
  readonly skillId: string;
  readonly masteryLevel: number; // 0.0 - 1.0
  readonly attemptsCount: number;
  readonly correctCount: number;
  readonly lastPracticed: Date | null;
  readonly confidenceTrend: ConfidenceTrend;
}

export interface AttemptRecord {
  readonly isCorrect: boolean;
  readonly timeSpentSeconds: number | null;
  readonly hintUsed: boolean;
  readonly tier: DifficultyLevel;
}

export interface SkillPriority {
  readonly skillId: string;
  readonly score: number;
  readonly reason: SkillPriorityReason;
}

export type SkillPriorityReason =
  | "prerequisite_gap"
  | "declining_confidence"
  | "stale"
  | "near_mastery"
  | "low_mastery"
  | "new_skill";

export interface DifficultyDecision {
  readonly tier: DifficultyLevel;
  readonly mode: SessionMode;
}

export interface SessionPacingState {
  readonly questionsInCurrentRun: number;
  readonly totalQuestions: number;
  readonly sessionStartTime: Date;
  readonly lastTeachingMoment: number; // question index of last teaching moment
  readonly recentAnswerTimesSeconds: readonly number[];
}

export interface PacingAction {
  readonly action: "continue_practice" | "insert_teaching" | "end_session" | "slow_down";
  readonly reason: string;
}

export interface MasteryUpdate {
  readonly newMasteryLevel: number;
  readonly newConfidenceTrend: ConfidenceTrend;
}

export interface MasteryWeightConfig {
  readonly weightRecent?: number;
  readonly weightOverall?: number;
  readonly weightTime?: number;
}

// ─── Tier Labels ─────────────────────────────────────────────────────

export const MASTERY_TIER_LABELS: Record<DifficultyLevel, string> = {
  1: "Getting Started",
  2: "Building Up",
  3: "In Progress",
  4: "Almost There",
  5: "Mastered",
};

export function tierLabel(tier: DifficultyLevel): string {
  return MASTERY_TIER_LABELS[tier];
}

// ─── Constants ────────────────────────────────────────────────────────

const MASTERY_THRESHOLD_LOW = 0.6;
const MASTERY_NEAR_LOW = 0.7;
const MASTERY_NEAR_HIGH = 0.85;
const STALE_DAYS = 7;
const STREAK_TO_ADVANCE = 3;
const STREAK_TO_DROP = 2;
const MAX_QUESTIONS_BEFORE_TEACHING = 5;
const SESSION_MIN_MINUTES = 25;
const SESSION_MAX_MINUTES = 35;
const RECENT_WINDOW = 10;
const SESSION_MAX_QUESTIONS = 15;
const RUSHING_THRESHOLD_SECONDS = 5;
const RUSHING_STREAK = 3;

// Mastery formula weights
const WEIGHT_RECENT = 0.7;
const WEIGHT_OVERALL = 0.2;
const WEIGHT_TIME = 0.1;

// Confidence multiplier: prevents high mastery from tiny sample sizes.
// Mastery is blended with MASTERY_PRIOR until MIN_ATTEMPTS_FOR_CONFIDENCE
// attempts are reached, at which point the formula fully trusts the data.
const MIN_ATTEMPTS_FOR_CONFIDENCE = 8;
const MASTERY_PRIOR = 0.3;

// Difficulty-weighted accuracy: harder questions count more toward mastery.
const TIER_ACCURACY_WEIGHT: Record<DifficultyLevel, number> = {
  1: 0.4,
  2: 0.6,
  3: 0.8,
  4: 1.0,
  5: 1.2,
};

// Priority scoring weights
// Ranking: prerequisite (100+) > declining (80-100) > low mastery (65-95)
//        > stale (40-80, capped) > near mastery (50-57) > new (35)
const PRIORITY_PREREQUISITE_GAP = 100;
const PRIORITY_DECLINING = 80;
const PRIORITY_STALE = 40;
const PRIORITY_STALE_CAP = 2; // cap staleness multiplier at 2x (was 3x)
const PRIORITY_NEAR_MASTERY = 50;
const PRIORITY_LOW_MASTERY = 65;
const PRIORITY_NEW_SKILL = 35;

// ─── 1. Skill Selection ──────────────────────────────────────────────

/**
 * Build a reverse dependency map: for each skill, which skills depend on it?
 * A skill that is a prerequisite for many others is higher-value to fix.
 */
function buildDependentsMap(
  skills: Map<string, Skill>
): Map<string, string[]> {
  const dependents = new Map<string, string[]>();
  for (const [, skill] of Array.from(skills.entries())) {
    for (const prereqId of skill.prerequisite_skills) {
      const existing = dependents.get(prereqId) ?? [];
      existing.push(skill.skill_id);
      dependents.set(prereqId, existing);
    }
  }
  return dependents;
}

/**
 * Score a single skill for priority selection.
 * Higher score = should be practiced sooner.
 */
function scoreSkill(
  skillId: string,
  state: StudentSkillState | undefined,
  dependents: Map<string, string[]>,
  now: Date
): SkillPriority {
  // Brand new skill — no data at all
  if (!state) {
    return { skillId, score: PRIORITY_NEW_SKILL, reason: "new_skill" };
  }

  // Skill with mastery data but no practice attempts (e.g. from diagnostic)
  // Treat as new only if mastery is also 0; otherwise use mastery for scoring
  if (state.attemptsCount === 0 && state.masteryLevel === 0) {
    return { skillId, score: PRIORITY_NEW_SKILL, reason: "new_skill" };
  }

  let bestScore = 0;
  let bestReason: SkillPriorityReason = "low_mastery";

  // Prerequisite gap: skill is below threshold AND has dependents
  const depCount = (dependents.get(skillId) ?? []).length;
  if (state.masteryLevel < MASTERY_THRESHOLD_LOW && depCount > 0) {
    const score = PRIORITY_PREREQUISITE_GAP + depCount * 5;
    if (score > bestScore) {
      bestScore = score;
      bestReason = "prerequisite_gap";
    }
  }

  // Declining confidence
  if (state.confidenceTrend === "declining") {
    const score = PRIORITY_DECLINING + (1 - state.masteryLevel) * 20;
    if (score > bestScore) {
      bestScore = score;
      bestReason = "declining_confidence";
    }
  }

  // Stale: not practiced in 7+ days
  // Mastery guard: highly mastered skills (>0.85) get a reduced stale boost,
  // tapering to zero at 1.0. SM-2 retention checks handle review of mastered skills.
  if (state.lastPracticed) {
    const daysSince = Math.floor(
      (now.getTime() - state.lastPracticed.getTime()) / (1000 * 60 * 60 * 24)
    );
    if (daysSince >= STALE_DAYS) {
      const staleness = Math.min(daysSince / STALE_DAYS, PRIORITY_STALE_CAP);
      const masteryDamper = state.masteryLevel >= MASTERY_NEAR_HIGH
        ? 1 - (state.masteryLevel - MASTERY_NEAR_HIGH) / (1 - MASTERY_NEAR_HIGH)
        : 1;
      const score = PRIORITY_STALE * staleness * masteryDamper;
      if (score > bestScore) {
        bestScore = score;
        bestReason = "stale";
      }
    }
  }

  // Near mastery: 0.7-0.85 — needs reinforcement
  if (
    state.masteryLevel >= MASTERY_NEAR_LOW &&
    state.masteryLevel <= MASTERY_NEAR_HIGH
  ) {
    const score =
      PRIORITY_NEAR_MASTERY + (MASTERY_NEAR_HIGH - state.masteryLevel) * 50;
    if (score > bestScore) {
      bestScore = score;
      bestReason = "near_mastery";
    }
  }

  // General low mastery (no dependents or doesn't qualify above)
  if (state.masteryLevel < MASTERY_THRESHOLD_LOW && bestScore === 0) {
    const score = PRIORITY_LOW_MASTERY + (1 - state.masteryLevel) * 30;
    if (score > bestScore) {
      bestScore = score;
      bestReason = "low_mastery";
    }
  }

  return {
    skillId,
    score: bestScore,
    reason: bestReason,
  };
}

/**
 * Select the next skill to practice from a domain.
 * Returns a priority-sorted list of skills (highest priority first).
 */
export function selectNextSkills(
  domainSkillIds: readonly string[],
  studentStates: ReadonlyMap<string, StudentSkillState>,
  now: Date = new Date()
): readonly SkillPriority[] {
  const allSkills = getAllSkills();
  const dependents = buildDependentsMap(allSkills);

  const priorities: SkillPriority[] = domainSkillIds.map((skillId) =>
    scoreSkill(skillId, studentStates.get(skillId), dependents, now)
  );

  return priorities.sort((a, b) => b.score - a.score);
}

// ─── 2. Difficulty Adjustment ────────────────────────────────────────

/**
 * Map mastery level (0.0-1.0) to a difficulty tier (1-5).
 */
export function masteryToTier(mastery: number): DifficultyLevel {
  if (mastery < 0.2) return 1;
  if (mastery < 0.4) return 2;
  if (mastery < 0.6) return 3;
  if (mastery < 0.8) return 4;
  return 5;
}

/**
 * Given recent attempts for a skill, decide the difficulty tier and mode.
 *
 * - 3 correct in a row → move up one tier
 * - 2 wrong in a row → drop one tier and switch to "teach" mode
 * - Otherwise stay at current tier in "practice" mode
 */
export function adjustDifficulty(
  currentMastery: number,
  recentAttempts: readonly AttemptRecord[]
): DifficultyDecision {
  const baseTier = masteryToTier(currentMastery);

  if (recentAttempts.length === 0) {
    return { tier: baseTier, mode: "practice" };
  }

  // For advancement: hint-assisted correct answers break the streak
  const independentAttempts = recentAttempts.map((a) => ({
    ...a,
    isCorrect: a.isCorrect && !a.hintUsed,
  }));
  const tailCorrectStreak = countTailStreak(independentAttempts, true);
  // For dropping: use original attempts (hints don't protect from wrong streak)
  const tailWrongStreak = countTailStreak(recentAttempts, false);

  if (tailWrongStreak >= STREAK_TO_DROP) {
    const droppedTier = Math.max(1, baseTier - 1) as DifficultyLevel;
    return { tier: droppedTier, mode: "teach" };
  }

  if (tailCorrectStreak >= STREAK_TO_ADVANCE) {
    const advancedTier = Math.min(5, baseTier + 1) as DifficultyLevel;
    return { tier: advancedTier, mode: "practice" };
  }

  return { tier: baseTier, mode: "practice" };
}

/**
 * Count consecutive matching results from the end of the array.
 */
function countTailStreak(
  attempts: readonly AttemptRecord[],
  matchCorrect: boolean
): number {
  let count = 0;
  for (let i = attempts.length - 1; i >= 0; i--) {
    if (attempts[i].isCorrect === matchCorrect) {
      count++;
    } else {
      break;
    }
  }
  return count;
}

// ─── 3. Session Pacing ───────────────────────────────────────────────

/**
 * Create initial pacing state for a new session.
 */
export function createPacingState(startTime: Date = new Date()): SessionPacingState {
  return {
    questionsInCurrentRun: 0,
    totalQuestions: 0,
    sessionStartTime: startTime,
    lastTeachingMoment: 0,
    recentAnswerTimesSeconds: [],
  };
}

/**
 * Determine whether to continue practice, insert a teaching moment,
 * or end the session.
 */
export function getNextPacingAction(
  state: SessionPacingState,
  now: Date = new Date()
): PacingAction {
  const elapsedMinutes =
    (now.getTime() - state.sessionStartTime.getTime()) / (1000 * 60);

  // Detect rushing: last N answers all under threshold
  const recentTimes = state.recentAnswerTimesSeconds.slice(-RUSHING_STREAK);
  if (
    recentTimes.length >= RUSHING_STREAK &&
    recentTimes.every((t) => t < RUSHING_THRESHOLD_SECONDS)
  ) {
    return {
      action: "slow_down",
      reason:
        "You're moving really fast! Take a moment to read each question carefully before answering — accuracy matters more than speed.",
    };
  }

  // Hard stop at max session time
  if (elapsedMinutes >= SESSION_MAX_MINUTES) {
    return {
      action: "end_session",
      reason: `Session reached ${SESSION_MAX_MINUTES} minutes — time for a break.`,
    };
  }

  // End session after reaching question limit
  if (state.totalQuestions >= SESSION_MAX_QUESTIONS) {
    return {
      action: "end_session",
      reason: "Great effort! You've answered a lot of questions — let's wrap up and see how you did.",
    };
  }

  // Suggest ending if past minimum and at a natural break
  if (
    elapsedMinutes >= SESSION_MIN_MINUTES &&
    state.questionsInCurrentRun >= 3
  ) {
    return {
      action: "end_session",
      reason:
        "Great work! We've been at it for a while — let's wrap up with a summary.",
    };
  }

  // Insert teaching after max consecutive questions
  if (state.questionsInCurrentRun >= MAX_QUESTIONS_BEFORE_TEACHING) {
    return {
      action: "insert_teaching",
      reason:
        "Time for a quick review! Let's make sure we understand the concepts before more practice.",
    };
  }

  return { action: "continue_practice", reason: "" };
}

/**
 * Advance pacing state after a practice question.
 */
export function advancePacingAfterQuestion(
  state: SessionPacingState,
  answerTimeSeconds?: number
): SessionPacingState {
  return {
    ...state,
    questionsInCurrentRun: state.questionsInCurrentRun + 1,
    totalQuestions: state.totalQuestions + 1,
    recentAnswerTimesSeconds:
      answerTimeSeconds !== undefined
        ? [...state.recentAnswerTimesSeconds, answerTimeSeconds]
        : state.recentAnswerTimesSeconds,
  };
}

/**
 * Reset the question run counter after a teaching moment.
 */
export function advancePacingAfterTeaching(
  state: SessionPacingState
): SessionPacingState {
  return {
    ...state,
    questionsInCurrentRun: 0,
    lastTeachingMoment: state.totalQuestions,
  };
}

// ─── 4. Mastery Update ───────────────────────────────────────────────

/**
 * Expected seconds per question by difficulty tier.
 * Used to compute time efficiency: finishing faster (but correctly) = higher score.
 */
const EXPECTED_SECONDS_BY_TIER: Record<DifficultyLevel, number> = {
  1: 30,
  2: 45,
  3: 60,
  4: 90,
  5: 120,
};

/**
 * Compute time efficiency score (0.0-1.0).
 * 1.0 = completed in expected time or less (while correct).
 * Scales down for slower attempts. Incorrect attempts score 0.
 */
function computeTimeEfficiency(
  attempts: readonly AttemptRecord[],
  tier: DifficultyLevel
): number {
  const expected = EXPECTED_SECONDS_BY_TIER[tier];
  const validAttempts = attempts.filter(
    (a) => a.timeSpentSeconds !== null && a.isCorrect
  );

  if (validAttempts.length === 0) return 0.5; // neutral default

  let totalScore = 0;
  for (const a of validAttempts) {
    const ratio = expected / Math.max(a.timeSpentSeconds!, 1);
    totalScore += Math.min(ratio, 1.0); // cap at 1.0, don't reward impossibly fast
  }

  return totalScore / validAttempts.length;
}

/**
 * Compute difficulty-weighted rolling accuracy over the last N attempts.
 *
 * Each attempt is weighted by its difficulty tier (TIER_ACCURACY_WEIGHT).
 * Correct answers earn the full tier weight (halved if hint-assisted).
 * The result is the ratio of earned weight to maximum possible weight,
 * so harder questions contribute more to (or detract more from) the score.
 */
function rollingAccuracy(
  attempts: readonly AttemptRecord[],
  window: number
): number {
  if (attempts.length === 0) return 0;
  const recent = attempts.slice(-window);
  let earned = 0;
  let maxPossible = 0;
  for (const a of recent) {
    const w = TIER_ACCURACY_WEIGHT[a.tier] ?? 0.8; // fallback to tier-3 weight
    maxPossible += w;
    if (a.isCorrect) {
      earned += a.hintUsed ? w * 0.5 : w; // scaffolded success discounted
    }
  }
  return maxPossible > 0 ? earned / maxPossible : 0;
}

/**
 * Determine the new confidence trend based on two accuracy snapshots.
 */
function computeTrend(
  recentAccuracy: number,
  overallAccuracy: number
): ConfidenceTrend {
  const delta = recentAccuracy - overallAccuracy;
  if (delta > 0.1) return "improving";
  if (delta < -0.1) return "declining";
  return "stable";
}

/**
 * Calculate updated mastery level and confidence trend after new attempts.
 *
 * Formula:
 *   raw = 0.7 * weighted_rolling_accuracy_last_10
 *       + 0.2 * weighted_overall_accuracy
 *       + 0.1 * time_efficiency_score
 *
 *   confidence = min(1, total_attempts / MIN_ATTEMPTS_FOR_CONFIDENCE)
 *   mastery    = confidence * raw + (1 - confidence) * MASTERY_PRIOR
 *
 * The confidence multiplier prevents tiny sample sizes (1-2 questions)
 * from producing artificially high mastery. The difficulty-weighted accuracy
 * ensures harder questions contribute more to the score.
 */
export function calculateMasteryUpdate(
  allAttempts: readonly AttemptRecord[],
  currentTier: DifficultyLevel,
  weights?: MasteryWeightConfig,
): MasteryUpdate {
  if (allAttempts.length === 0) {
    return { newMasteryLevel: 0, newConfidenceTrend: "stable" };
  }

  const wR = weights?.weightRecent ?? WEIGHT_RECENT;
  const wO = weights?.weightOverall ?? WEIGHT_OVERALL;
  const wT = weights?.weightTime ?? WEIGHT_TIME;

  const recentAcc = rollingAccuracy(allAttempts, RECENT_WINDOW);
  const overallAcc = rollingAccuracy(allAttempts, allAttempts.length);
  const timeEff = computeTimeEfficiency(
    allAttempts.slice(-RECENT_WINDOW),
    currentTier
  );

  const rawMastery = wR * recentAcc + wO * overallAcc + wT * timeEff;

  // Confidence multiplier: blend raw score with a conservative prior
  // until enough attempts have been collected to trust the data.
  const confidence = Math.min(1, allAttempts.length / MIN_ATTEMPTS_FOR_CONFIDENCE);
  const blended = confidence * rawMastery + (1 - confidence) * MASTERY_PRIOR;

  const newMasteryLevel = Math.max(0, Math.min(1, blended));
  const newConfidenceTrend = computeTrend(recentAcc, overallAcc);

  return {
    newMasteryLevel: Math.round(newMasteryLevel * 1000) / 1000, // 3 decimal places
    newConfidenceTrend,
  };
}

File: python_app/app/lib/adaptive.py
"""
Adaptive learning engine — ported from TypeScript src/lib/adaptive.ts.

Handles:
1. Skill selection and priority scoring
2. Difficulty adjustment (tier up/down based on streaks)
3. Session pacing (when to teach, end, slow down)
4. Mastery calculation (weighted rolling accuracy + time efficiency)
"""
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import datetime, UTC
from typing import Literal, Optional
from app.lib.curriculum import get_all_skills

# ─── Types ────────────────────────────────────────────────────────────

ConfidenceTrend = Literal["improving", "stable", "declining"]
SessionMode = Literal["teach", "practice"]
DifficultyLevel = Literal[1, 2, 3, 4, 5]
SkillPriorityReason = Literal[
    "prerequisite_gap", "declining_confidence", "stale",
    "near_mastery", "low_mastery", "new_skill"
]
PacingActionType = Literal["continue_practice", "insert_teaching", "end_session", "slow_down"]


@dataclass(frozen=True)
class StudentSkillState:
    skill_id: str
    mastery_level: float  # 0.0 - 1.0
    attempts_count: int
    correct_count: int
    last_practiced: Optional[datetime]
    confidence_trend: ConfidenceTrend


@dataclass(frozen=True)
class AttemptRecord:
    is_correct: bool
    time_spent_seconds: Optional[float]
    hint_used: bool
    tier: int  # 1-5


@dataclass(frozen=True)
class SkillPriority:
    skill_id: str
    score: float
    reason: SkillPriorityReason


@dataclass(frozen=True)
class DifficultyDecision:
    tier: int  # 1-5
    mode: SessionMode


@dataclass
class SessionPacingState:
    questions_in_current_run: int = 0
    total_questions: int = 0
    session_start_time: datetime = field(default_factory=lambda: datetime.now(UTC).replace(tzinfo=None))
    last_teaching_moment: int = 0
    recent_answer_times_seconds: list[float] = field(default_factory=list)


@dataclass(frozen=True)
class PacingAction:
    action: PacingActionType
    reason: str


@dataclass(frozen=True)
class MasteryUpdate:
    new_mastery_level: float
    new_confidence_trend: ConfidenceTrend


# ─── Tier Labels ─────────────────────────────────────────────────────

MASTERY_TIER_LABELS: dict[int, str] = {
    1: "Getting Started",
    2: "Building Up",
    3: "In Progress",
    4: "Almost There",
    5: "Mastered",
}


def tier_label(tier: int) -> str:
    return MASTERY_TIER_LABELS.get(tier, "Unknown")


# ─── Constants ────────────────────────────────────────────────────────

MASTERY_THRESHOLD_LOW = 0.6
MASTERY_NEAR_LOW = 0.7
MASTERY_NEAR_HIGH = 0.85
STALE_DAYS = 7
STREAK_TO_ADVANCE = 3
STREAK_TO_DROP = 2
MAX_QUESTIONS_BEFORE_TEACHING = 5
SESSION_MIN_MINUTES = 25
SESSION_MAX_MINUTES = 35
RECENT_WINDOW = 10
SESSION_MAX_QUESTIONS = 15
RUSHING_THRESHOLD_SECONDS = 5
RUSHING_STREAK = 3

WEIGHT_RECENT = 0.7
WEIGHT_OVERALL = 0.2
WEIGHT_TIME = 0.1

MIN_ATTEMPTS_FOR_CONFIDENCE = 8
MASTERY_PRIOR = 0.3

TIER_ACCURACY_WEIGHT: dict[int, float] = {
    1: 0.4,
    2: 0.6,
    3: 0.8,
    4: 1.0,
    5: 1.2,
}

PRIORITY_PREREQUISITE_GAP = 100
PRIORITY_DECLINING = 80
PRIORITY_STALE = 40
PRIORITY_STALE_CAP = 2
PRIORITY_NEAR_MASTERY = 50
PRIORITY_LOW_MASTERY = 65
PRIORITY_NEW_SKILL = 35

EXPECTED_SECONDS_BY_TIER: dict[int, float] = {
    1: 30,
    2: 45,
    3: 60,
    4: 90,
    5: 120,
}


# ─── 1. Skill Selection ──────────────────────────────────────────────

def _build_dependents_map(skills: dict[str, dict]) -> dict[str, list[str]]:
    """Build reverse dependency map: prereq_id -> [skill_ids that depend on it]."""
    dependents: dict[str, list[str]] = {}
    for skill_id, skill in skills.items():
        for prereq_id in skill.get("prerequisite_skills", []):
            dependents.setdefault(prereq_id, []).append(skill_id)
    return dependents


def _score_skill(
    skill_id: str,
    state: Optional[StudentSkillState],
    dependents: dict[str, list[str]],
    now: datetime,
) -> SkillPriority:
    """Compute priority score for a single skill."""
    if state is None or (state.attempts_count == 0 and state.mastery_level == 0):
        return SkillPriority(skill_id=skill_id, score=PRIORITY_NEW_SKILL, reason="new_skill")

    best_score: float = 0.0
    best_reason: SkillPriorityReason = "low_mastery"

    # Prerequisite gap
    dep_count = len(dependents.get(skill_id, []))
    if state.mastery_level < MASTERY_THRESHOLD_LOW and dep_count > 0:
        score = PRIORITY_PREREQUISITE_GAP + dep_count * 5
        if score > best_score:
            best_score = score
            best_reason = "prerequisite_gap"

    # Declining confidence
    if state.confidence_trend == "declining":
        score = PRIORITY_DECLINING + (1 - state.mastery_level) * 20
        if score > best_score:
            best_score = score
            best_reason = "declining_confidence"

    # Stale
    if state.last_practiced:
        lp = state.last_practiced
        if lp.tzinfo is None:
            lp = lp.replace(tzinfo=None)
        now_naive = now.replace(tzinfo=None) if now.tzinfo else now
        days_since = (now_naive - lp).days
        if days_since >= STALE_DAYS:
            staleness = min(days_since / STALE_DAYS, PRIORITY_STALE_CAP)
            mastery_damper = (
                1 - (state.mastery_level - MASTERY_NEAR_HIGH) / (1 - MASTERY_NEAR_HIGH)
                if state.mastery_level >= MASTERY_NEAR_HIGH
                else 1.0
            )
            score = PRIORITY_STALE * staleness * mastery_damper
            if score > best_score:
                best_score = score
                best_reason = "stale"

    # Near mastery
    if MASTERY_NEAR_LOW <= state.mastery_level <= MASTERY_NEAR_HIGH:
        score = PRIORITY_NEAR_MASTERY + (MASTERY_NEAR_HIGH - state.mastery_level) * 50
        if score > best_score:
            best_score = score
            best_reason = "near_mastery"

    # General low mastery
    if state.mastery_level < MASTERY_THRESHOLD_LOW and best_score == 0:
        score = PRIORITY_LOW_MASTERY + (1 - state.mastery_level) * 30
        if score > best_score:
            best_score = score
            best_reason = "low_mastery"

    return SkillPriority(skill_id=skill_id, score=best_score, reason=best_reason)


def select_next_skills(
    domain_skill_ids: list[str],
    student_states: dict[str, StudentSkillState],
    now: Optional[datetime] = None,
) -> list[SkillPriority]:
    """Return priority-sorted skills (highest first) for the given domain."""
    if now is None:
        now = datetime.now(UTC).replace(tzinfo=None)
    all_skills = get_all_skills()
    dependents = _build_dependents_map(all_skills)
    priorities = [
        _score_skill(skill_id, student_states.get(skill_id), dependents, now)
        for skill_id in domain_skill_ids
    ]
    return sorted(priorities, key=lambda p: p.score, reverse=True)


# ─── 2. Difficulty Adjustment ────────────────────────────────────────

def mastery_to_tier(mastery: float) -> int:
    """Map mastery level [0,1] to difficulty tier [1,5]."""
    if mastery < 0.2:
        return 1
    if mastery < 0.4:
        return 2
    if mastery < 0.6:
        return 3
    if mastery < 0.8:
        return 4
    return 5


def _count_tail_streak(attempts: list[AttemptRecord], match_correct: bool) -> int:
    """Count consecutive matching results from the end of the list."""
    count = 0
    for attempt in reversed(attempts):
        if attempt.is_correct == match_correct:
            count += 1
        else:
            break
    return count


def adjust_difficulty(
    current_mastery: float,
    recent_attempts: list[AttemptRecord],
) -> DifficultyDecision:
    """Determine tier and mode based on recent attempt streak."""
    base_tier = mastery_to_tier(current_mastery)

    if not recent_attempts:
        return DifficultyDecision(tier=base_tier, mode="practice")

    # For advancement: hint-assisted correct don't count
    independent_attempts = [
        AttemptRecord(
            is_correct=a.is_correct and not a.hint_used,
            time_spent_seconds=a.time_spent_seconds,
            hint_used=a.hint_used,
            tier=a.tier,
        )
        for a in recent_attempts
    ]

    tail_correct_streak = _count_tail_streak(independent_attempts, True)
    tail_wrong_streak = _count_tail_streak(recent_attempts, False)

    if tail_wrong_streak >= STREAK_TO_DROP:
        dropped_tier = max(1, base_tier - 1)
        return DifficultyDecision(tier=dropped_tier, mode="teach")

    if tail_correct_streak >= STREAK_TO_ADVANCE:
        advanced_tier = min(5, base_tier + 1)
        return DifficultyDecision(tier=advanced_tier, mode="practice")

    return DifficultyDecision(tier=base_tier, mode="practice")


# ─── 3. Session Pacing ───────────────────────────────────────────────

def create_pacing_state(start_time: Optional[datetime] = None) -> SessionPacingState:
    """Create initial pacing state."""
    if start_time is None:
        start_time = datetime.now(UTC).replace(tzinfo=None)
    return SessionPacingState(session_start_time=start_time)


def get_next_pacing_action(
    state: SessionPacingState,
    now: Optional[datetime] = None,
) -> PacingAction:
    """Determine the next pacing action."""
    if now is None:
        now = datetime.now(UTC).replace(tzinfo=None)

    start = state.session_start_time
    if start.tzinfo:
        start = start.replace(tzinfo=None)
    now_naive = now.replace(tzinfo=None) if now.tzinfo else now

    elapsed_minutes = (now_naive - start).total_seconds() / 60

    # Detect rushing
    recent_times = state.recent_answer_times_seconds[-RUSHING_STREAK:]
    if len(recent_times) >= RUSHING_STREAK and all(t < RUSHING_THRESHOLD_SECONDS for t in recent_times):
        return PacingAction(
            action="slow_down",
            reason=(
                "You're moving really fast! Take a moment to read each question "
                "carefully before answering — accuracy matters more than speed."
            ),
        )

    if elapsed_minutes >= SESSION_MAX_MINUTES:
        return PacingAction(
            action="end_session",
            reason=f"Session reached {SESSION_MAX_MINUTES} minutes — time for a break.",
        )

    if state.total_questions >= SESSION_MAX_QUESTIONS:
        return PacingAction(
            action="end_session",
            reason="Great effort! You've answered a lot of questions — let's wrap up and see how you did.",
        )

    if elapsed_minutes >= SESSION_MIN_MINUTES and state.questions_in_current_run >= 3:
        return PacingAction(
            action="end_session",
            reason="Great work! We've been at it for a while — let's wrap up with a summary.",
        )

    if state.questions_in_current_run >= MAX_QUESTIONS_BEFORE_TEACHING:
        return PacingAction(
            action="insert_teaching",
            reason="Time for a quick review! Let's make sure we understand the concepts before more practice.",
        )

    return PacingAction(action="continue_practice", reason="")


def advance_pacing_after_question(
    state: SessionPacingState,
    answer_time_seconds: Optional[float] = None,
) -> SessionPacingState:
    """Return new state after a practice question."""
    new_times = list(state.recent_answer_times_seconds)
    if answer_time_seconds is not None:
        new_times.append(answer_time_seconds)
    return SessionPacingState(
        questions_in_current_run=state.questions_in_current_run + 1,
        total_questions=state.total_questions + 1,
        session_start_time=state.session_start_time,
        last_teaching_moment=state.last_teaching_moment,
        recent_answer_times_seconds=new_times,
    )


def advance_pacing_after_teaching(state: SessionPacingState) -> SessionPacingState:
    """Reset question run counter after a teaching moment."""
    return SessionPacingState(
        questions_in_current_run=0,
        total_questions=state.total_questions,
        session_start_time=state.session_start_time,
        last_teaching_moment=state.total_questions,
        recent_answer_times_seconds=state.recent_answer_times_seconds,
    )


# ─── 4. Mastery Update ───────────────────────────────────────────────

def _compute_time_efficiency(attempts: list[AttemptRecord], tier: int) -> float:
    """Compute time efficiency score [0,1]."""
    expected = EXPECTED_SECONDS_BY_TIER.get(tier, 60)
    valid = [a for a in attempts if a.time_spent_seconds is not None and a.is_correct]
    if not valid:
        return 0.5  # neutral default

    total = sum(min(expected / max(a.time_spent_seconds, 1), 1.0) for a in valid)
    return total / len(valid)


def _rolling_accuracy(attempts: list[AttemptRecord], window: int) -> float:
    """Compute difficulty-weighted rolling accuracy over last `window` attempts."""
    if not attempts:
        return 0.0
    recent = attempts[-window:]
    earned = 0.0
    max_possible = 0.0
    for a in recent:
        w = TIER_ACCURACY_WEIGHT.get(a.tier, 0.8)
        max_possible += w
        if a.is_correct:
            earned += w * 0.5 if a.hint_used else w
    return earned / max_possible if max_possible > 0 else 0.0


def _compute_trend(recent_acc: float, overall_acc: float) -> ConfidenceTrend:
    delta = recent_acc - overall_acc
    if delta > 0.1:
        return "improving"
    if delta < -0.1:
        return "declining"
    return "stable"


def calculate_mastery_update(
    all_attempts: list[AttemptRecord],
    current_tier: int,
    weight_recent: float = WEIGHT_RECENT,
    weight_overall: float = WEIGHT_OVERALL,
    weight_time: float = WEIGHT_TIME,
) -> MasteryUpdate:
    """
    Calculate new mastery level and confidence trend.

    Formula:
        raw = 0.7 * weighted_rolling_accuracy_last_10
            + 0.2 * weighted_overall_accuracy
            + 0.1 * time_efficiency_score

        confidence = min(1, total_attempts / MIN_ATTEMPTS_FOR_CONFIDENCE)
        mastery = confidence * raw + (1 - confidence) * MASTERY_PRIOR
    """
    if not all_attempts:
        return MasteryUpdate(new_mastery_level=0.0, new_confidence_trend="stable")

    recent_acc = _rolling_accuracy(all_attempts, RECENT_WINDOW)
    overall_acc = _rolling_accuracy(all_attempts, len(all_attempts))
    time_eff = _compute_time_efficiency(all_attempts[-RECENT_WINDOW:], current_tier)

    raw = weight_recent * recent_acc + weight_overall * overall_acc + weight_time * time_eff

    confidence = min(1.0, len(all_attempts) / MIN_ATTEMPTS_FOR_CONFIDENCE)
    blended = confidence * raw + (1 - confidence) * MASTERY_PRIOR

    new_mastery = max(0.0, min(1.0, blended))
    new_mastery = round(new_mastery, 3)
    trend = _compute_trend(recent_acc, overall_acc)

    return MasteryUpdate(new_mastery_level=new_mastery, new_confidence_trend=trend)
